File: ios/demo-objc/api_demo/ViewController.m
#import "ViewController.h"
#import "AppDelegate.h"

@interface ViewController ()

@property(nonatomic, assign) UIInterfaceOrientation statusBarOrientation;

@end

@implementation ViewController{
	NSMutableDictionary* mediaStreams;
	NSString* roomName;
}

- (void)viewDidLoad {
	[super viewDidLoad];
	
	self.statusBarOrientation = [UIApplication sharedApplication].statusBarOrientation;
	
	self.roomInput.delegate = self;
	[self.roomInput becomeFirstResponder];
	
	self.view.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"wood.png"]];
	
	mediaStreams = [[NSMutableDictionary alloc] init];
}

- (void)viewDidAppear:(BOOL)animated {
	[super viewDidAppear:animated];
	
	[self resizeVideoViews];
	
	UIApplication* myApp = [UIApplication sharedApplication];
	myApp.idleTimerDisabled = YES;
	
}

- (void)addMediaStream: (MediaStream*)mediaStream {
	NSLog(@"addMediaStream pid:%@", [ [mediaStream getPeerStream] getId ] );
	
	[mediaStreams setObject:mediaStream forKey:[ [mediaStream getPeerStream] getId ] ];
	mediaStream.delegate = self;
	
	[self resizeVideoViews];
	
	UIView* v = [ [ mediaStream getVideoView ] getView];
	[self.videosView addSubview: v ];
}

- (void)removePeerStream: (PeerStream*)peerStream {
	
	NSString* id = [peerStream getId];
	
	[ [ [ [mediaStreams valueForKey:id] getVideoView ] getView] removeFromSuperview ];
	
	[mediaStreams removeObjectForKey:id];
	
	[self resizeVideoViews];
}

- (void)sizeVideoView:(VideoView*) vv ratio:(float)ratio top:(int)top left:(int)left width:(int)maxWidth height:(int)maxHeight{
	
	int height = maxHeight;
	int width = height * ratio;
	if ( width > maxWidth) {
		width = maxWidth;
		height = maxWidth / ratio;
	}
	
	int centeredTop = top + (maxHeight-height) /2;
	int centeredLeft = left + (maxWidth-width) /2;
	
	UIView* view = [vv getView];
	
	CGRect frame = view.frame;
	
	frame.size.height = height;
	frame.size.width = width;
	frame.origin.x = centeredLeft;
	frame.origin.y = centeredTop;
	
	view.frame = frame;
}

- (void)resizeVideoViews {
	
	CGRect videosRect = [self.videosView frame];
	CGFloat videosWidth = videosRect.size.width;
	CGFloat videosHeight = videosRect.size.height;
	NSArray* keys = [mediaStreams allKeys];
	NSUInteger count = keys.count;
	
	if ( !count ) return;
	
	int maxHeight = videosHeight / ( count/2 + count%2 );
	int maxWidth = videosWidth / ( (count>1) ? 2 : 1 );
	
	for (int i=0; i<count; i++) {
		NSString* id = keys[i];
		MediaStream* ms = [ mediaStreams valueForKey:id ];
		
		[self sizeVideoView:[ms getVideoView] ratio:[ms getVideoRatio] top:(i/2)*maxHeight left:(i%2)*maxWidth width:maxWidth height:maxHeight ];
	}
	
	// Be sure that leave button is visible
	[[self.leaveBtn superview] bringSubviewToFront:self.leaveBtn];
}

- (void)onVideoRatioChange:(float)ratio media:(MediaStream *)mediaStream {
	[self resizeVideoViews];
}

- (void)onVideoSizeChange:(int)width height:(int)height media:(MediaStream*)mediaStream {
	// Can be used to show video only at this moment
	NSLog(@"onVideoSizeChange");
}

- (void)onVideoReady:(MediaStream *)mediaStream {
	// Can be used to show video only at this moment
	NSLog(@"videoReady");
}

- (void)resetUI {
	[self.roomInput resignFirstResponder];
	self.roomInput.text = nil;
	[self inRoom:NO];
}

#pragma mark - UITextFieldDelegate
- (void)textFieldDidEndEditing:(UITextField*)textField {
	[ self join ];
}


- (BOOL)textFieldShouldReturn:(UITextField*)textField {
	[textField resignFirstResponder];
	return YES;
}

-(NSUInteger)supportedInterfaceOrientations {
	return UIInterfaceOrientationMaskAll;
}

- (void)viewDidLayoutSubviews {
	if (self.statusBarOrientation != [UIApplication sharedApplication].statusBarOrientation) {
		self.statusBarOrientation = [UIApplication sharedApplication].statusBarOrientation;
		[[NSNotificationCenter defaultCenter]
		 postNotificationName:@"StatusBarOrientationDidChange"
		 object:nil ];
		
		[ self resizeVideoViews ];
	}
}

- (IBAction)leaveBtn:(id)sender {
	AppDelegate * appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
	[appDelegate leave:roomName];
}

- (void) join {
	roomName = self.roomInput.text;
	if ([roomName length] == 0) {
		return;
	}
	
	[self.roomInput resignFirstResponder];
	
	AppDelegate * appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
	[appDelegate join:roomName];
}

- (IBAction)joinBtn:(id)sender {
	[ self join ];
}

- (void)inRoom:(BOOL)inside{
	self.roomInput.hidden = inside;
	self.videosView.hidden = !inside;
	self.leaveBtn.hidden = !inside;
	self.joinBtn.hidden = inside;
	self.statusLabel.hidden = inside;
}

- (void)updateConnectionStatus: (ConferenceConnection)status {
	if ( status == CONNECTED ) {
		[self.joinBtn setEnabled:YES];
		self.joinBtn.userInteractionEnabled = YES;
	} else {
		[self.joinBtn setEnabled:NO];
		self.joinBtn.userInteractionEnabled = NO;
	}
	switch ( status ) {
		case DISCONNECTED:
			self.statusLabel.text = @"Disconnected";
			break;
		case CONNECTING:
		case CONNECTING_SENDREQUEST:
			self.statusLabel.text = @"Connecting";
			break;
		case CONNECTED:
			self.statusLabel.text = @"Connected";
			break;
	}
}

@end

File: demo-swift/demo/ViewController.swift
//
//  ViewController.swift
//  demo
//
//  Created by Aurélien Hiron on 11/04/2023.
//

import UIKit
import bistriAPI2

class ViewController: UIViewController, UITextFieldDelegate, MediaStreamDelegate {
	
	@IBOutlet var contentView: UIView!
	@IBOutlet weak var roomInput: UITextField!
	@IBOutlet weak var videosView: UIView!
	@IBOutlet weak var statusLabel: UILabel!
	@IBOutlet weak var leaveBtn: UIButton!
	@IBOutlet weak var joinBtn: UIButton!

	var orientation: UIInterfaceOrientation = .unknown
	var mediaStreams: [String: MediaStream] = [:]
	var roomName: String?

	override func viewDidLoad() {
		super.viewDidLoad()
		orientation = getInterfaceOrientation()
		
		roomInput.delegate = self
		roomInput.becomeFirstResponder()

		mediaStreams = [:]
	}
	
	func getInterfaceOrientation() -> UIInterfaceOrientation {
		let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene
		return windowScene?.interfaceOrientation ?? UIInterfaceOrientation.unknown
	}

	override func viewDidAppear(_ animated: Bool) {
		super.viewDidAppear(animated)

		resizeVideoViews()

		let myApp = UIApplication.shared
		myApp.isIdleTimerDisabled = true
	}

	func addMediaStream(_ mediaStream: MediaStream) {
		print("addMediaStream pid:\(mediaStream.getPeerStream().getId())")

		mediaStreams[mediaStream.getPeerStream().getId()] = mediaStream
		mediaStream.delegate = self

		let v = mediaStream.getVideoView().getView()
		videosView.addSubview(v!)

		resizeVideoViews()

	}

	func removePeerStream(_ peerStream: PeerStream) {
		guard let id = peerStream.getId() else { return }

		mediaStreams[id]?.getVideoView().getView().removeFromSuperview()
		mediaStreams.removeValue(forKey: id)

		resizeVideoViews()
	}

	func sizeVideoView(_ vv: VideoView, ratio: Float, top: Int, left: Int, maxWidth: Int, maxHeight: Int) {
		
		var height = maxHeight
		var width = Int(Float(height) * ratio)
		if width > maxWidth {
			width = maxWidth
			height = Int(Float(width) / ratio)
		}

		let centeredTop = top + (maxHeight - height) / 2
		let centeredLeft = left + (maxWidth - width) / 2

		guard let view = vv.getView() else { return }
		var frame = view.frame
		frame.size.height = CGFloat(height)
		frame.size.width = CGFloat(width)
		frame.origin.x = CGFloat(centeredLeft)
		frame.origin.y = CGFloat(centeredTop)

		view.frame = frame
	}

	func resizeVideoViews() {

		let videosRect = videosView.frame
		let videosWidth = videosRect.size.width
		let videosHeight = videosRect.size.height
		let keys = Array(mediaStreams.keys)
		let count = keys.count

		if count == 0 {
			return
		}

		let maxHeight = Int(videosHeight / CGFloat(count/2 + count%2))
		let maxWidth = Int(videosWidth / CGFloat(count > 1 ? 2 : 1))

		for i in 0..<count {
			let id = keys[i]
			let ms = mediaStreams[id]

			sizeVideoView(ms!.getVideoView(), ratio: ms!.getVideoRatio(), top: (i/2)*maxHeight, left: (i%2)*maxWidth, maxWidth: maxWidth, maxHeight: maxHeight)
		}

		// Be sure that leave button is visible
		leaveBtn.superview?.bringSubviewToFront(leaveBtn)
	}

	func onVideoRatioChange(_ ratio: Float, media mediaStream: MediaStream) {
		resizeVideoViews()
	}

	func onVideoSizeChange(_ width: Int32, height: Int32, media mediaStream: MediaStream!) {
		// Peut être utilisé pour afficher uniquement la vidéo à ce moment-là
		print("onVideoSizeChange")
	}

	func onVideoReady(_ mediaStream: MediaStream) {
		// Peut être utilisé pour afficher uniquement la vidéo à ce moment-là
		print("videoReady")
	}

	func resetUI() {
		roomInput.resignFirstResponder()
		roomInput.text = nil
		inRoom(false)
	}

	// MARK: - UITextFieldDelegate

	func textFieldDidEndEditing(_ textField: UITextField) {
		join()
	}

	func textFieldShouldReturn(_ textField: UITextField) -> Bool {
		textField.resignFirstResponder()
		return true
	}

	override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
		return .all
	}

	override func viewDidLayoutSubviews() {
		let newOrientation = getInterfaceOrientation()
		if orientation != newOrientation {
			orientation = newOrientation
			NotificationCenter.default.post(name: NSNotification.Name("StatusBarOrientationDidChange"), object: nil)
			resizeVideoViews()
		}
	}

	@IBAction func leaveBtnTapped(_ sender: Any) {
		guard let roomName = roomName else {return}
		let appDelegate = UIApplication.shared.delegate as? AppDelegate
		appDelegate?.leave(roomName)
	}

	func join() {
		roomName = roomInput.text
		guard let roomName = roomName else {return}

		roomInput.resignFirstResponder()

		let appDelegate = UIApplication.shared.delegate as? AppDelegate
		appDelegate?.join(roomName)
	}

	@IBAction func joinBtnTapped(_ sender: Any) {
		print("join tapped")
		join()
	}

	func inRoom(_ inside: Bool) {
		roomInput.isHidden = inside
		videosView.isHidden = !inside
		leaveBtn.isHidden = !inside
		joinBtn.isHidden = inside
		statusLabel.isHidden = inside
	}

	func updateConnectionStatus(_ status: ConferenceConnection) {
		let connected = (status == CONNECTED)
		joinBtn.isEnabled = connected
		joinBtn.isUserInteractionEnabled = connected

		switch status {
		case CONNECTING:
			statusLabel.text = "Connecting"
		case CONNECTING_SENDREQUEST:
			statusLabel.text = "Connecting"
		case CONNECTED:
			statusLabel.text = "Connected"
		default:
			statusLabel.text = "Disconnected"
		}
	}

	// Autres méthodes et fonctionnalités de la classe ViewController en Swift
}




